File: scala/src/main/scala/hearth/crypto/Ecvrf.scala
package hearth.crypto

/** ECVRF-EDWARDS25519-SHA512-TAI, the Ed25519 verifiable random function from
  * RFC 9381 (suite_string = 0x03).
  *
  * It reuses the exact Ed25519 key derived from BIP-39/SLIP-0010: the VRF secret
  * scalar is `clamp(SHA-512(seed))` and the VRF public key is the Ed25519 public
  * key. So a single Ledger-derived key both signs (EdDSA) and produces VRF
  * proofs for miner election. Curve/scalar arithmetic runs on a [[CryptoBackend]]
  * (libsodium or pure-JVM), defaulting to [[Crypto]].
  */
object Ecvrf:

  private val Suite: Byte = 0x03
  private val PtLen  = Crypto.PointBytes  // 32
  private val CLen   = 16                 // challenge length in bytes
  private val ProofLen = PtLen + CLen + PtLen // 80
  private val Identity: Array[Byte] = Array[Byte](1) ++ Array.fill(31)(0.toByte)

  /** A VRF proof pi = Gamma(32) || c(16) || s(32). */
  final case class Proof(gamma: Array[Byte], c: Array[Byte], s: Array[Byte]):
    def bytes: Array[Byte] = gamma ++ c ++ s

  /** Prove: returns the proof pi and the VRF output beta (64 bytes). */
  def prove(seed: Array[Byte], alpha: Array[Byte])(using b: CryptoBackend): (Proof, Array[Byte]) =
    val x  = Ed25519.secretScalar(seed)
    val y  = b.scalarmultBaseNoclamp(x) // public key Y = x*B
    val h  = encodeToCurve(y, alpha)
    val gamma = b.scalarmultNoclamp(x, h).getOrElse(sys.error("Gamma = x*H failed"))
    val k  = nonce(seed, h)
    val u  = b.scalarmultBaseNoclamp(k)                        // U = k*B
    val v  = b.scalarmultNoclamp(k, h).getOrElse(sys.error("V = k*H failed")) // V = k*H
    val c  = challenge(y, h, gamma, u, v)                      // 16 bytes
    val c32 = c ++ Array.fill(PtLen - CLen)(0.toByte)
    val s  = b.scalarAdd(k, b.scalarMul(c32, x))               // s = k + c*x mod L
    val proof = Proof(gamma, c, s)
    (proof, proofToHash(proof))

  /** Verify a proof against the public key and alpha. Returns Some(beta) if
    * valid, None otherwise.
    */
  def verify(publicKey: Array[Byte], alpha: Array[Byte], pi: Array[Byte])(using b: CryptoBackend): Option[Array[Byte]] =
    decode(pi).flatMap { proof =>
      val h   = encodeToCurve(publicKey, alpha)
      val c32 = proof.c ++ Array.fill(PtLen - CLen)(0.toByte)
      for
        sB     <- Some(b.scalarmultBaseNoclamp(proof.s))
        cY     <- b.scalarmultNoclamp(c32, publicKey)
        u      <- b.pointSub(sB, cY)                       // U = s*B - c*Y
        sH     <- b.scalarmultNoclamp(proof.s, h)
        cGamma <- b.scalarmultNoclamp(c32, proof.gamma)
        v      <- b.pointSub(sH, cGamma)                   // V = s*H - c*Gamma
        cPrime  = challenge(publicKey, h, proof.gamma, u, v)
        if java.util.Arrays.equals(cPrime, proof.c)
      yield proofToHash(proof)
    }

  /** proof_to_hash: beta = SHA-512(suite || 0x03 || point_to_string(8*Gamma) || 0x00). */
  def proofToHash(proof: Proof)(using b: CryptoBackend): Array[Byte] =
    val gamma8 = cofactorClear(proof.gamma).getOrElse(sys.error("8*Gamma failed"))
    b.sha512(Array(Suite, 0x03.toByte) ++ gamma8 ++ Array(0x00.toByte))

  def decode(pi: Array[Byte]): Option[Proof] =
    if pi.length != ProofLen then None
    else Some(Proof(pi.slice(0, PtLen), pi.slice(PtLen, PtLen + CLen), pi.slice(PtLen + CLen, ProofLen)))

  // --- Internals -----------------------------------------------------------

  /** ECVRF_encode_to_curve_try_and_increment (RFC 9381 §5.4.1.1). */
  private def encodeToCurve(pkString: Array[Byte], alpha: Array[Byte])(using b: CryptoBackend): Array[Byte] =
    var ctr = 0
    var result: Option[Array[Byte]] = None
    while result.isEmpty do
      if ctr > 255 then sys.error("encode_to_curve: no valid point found")
      val hashString = b.sha512(
        Array(Suite, 0x01.toByte) ++ pkString ++ alpha ++ Array(ctr.toByte) ++ Array(0x00.toByte)
      )
      val candidate = hashString.slice(0, PtLen) // string_to_point(first 32 bytes)
      cofactorClear(candidate) match             // INVALID (off-curve) => None => next ctr
        case Some(hp) if !java.util.Arrays.equals(hp, Identity) => result = Some(hp)
        case _                                                   => ()
      ctr += 1
    result.get

  /** Multiply a compressed point by the cofactor 8 via three doublings. Uses
    * point addition, which only requires the point to be on the curve (not on
    * the prime-order subgroup) — exactly what try-and-increment needs. Returns
    * None if the input does not decode to a curve point.
    */
  private def cofactorClear(p: Array[Byte])(using b: CryptoBackend): Option[Array[Byte]] =
    for
      p2 <- b.pointAdd(p, p)
      p4 <- b.pointAdd(p2, p2)
      p8 <- b.pointAdd(p4, p4)
    yield p8

  /** ECVRF_nonce_generation_RFC8032 (RFC 9381 §5.4.2.2). */
  private def nonce(seed: Array[Byte], hString: Array[Byte])(using b: CryptoBackend): Array[Byte] =
    val kString = b.sha512(Ed25519.noncePrefix(seed) ++ hString)
    b.scalarReduce(kString) // k = string_to_int(k_string) mod L

  /** ECVRF_challenge_generation (RFC 9381 §5.4.3): first 16 bytes of
    * SHA-512(suite || 0x02 || Y || H || Gamma || U || V || 0x00).
    */
  private def challenge(y: Array[Byte], h: Array[Byte], gamma: Array[Byte], u: Array[Byte], v: Array[Byte])(using b: CryptoBackend): Array[Byte] =
    b.sha512(Array(Suite, 0x02.toByte) ++ y ++ h ++ gamma ++ u ++ v ++ Array(0x00.toByte)).slice(0, CLen)

File: java/src/main/java/tech/hearth/crypto/Ecvrf.java
package tech.hearth.crypto;

import java.util.Arrays;
import java.util.Optional;

/**
 * ECVRF-EDWARDS25519-SHA512-TAI, the Ed25519 verifiable random function from
 * RFC 9381 (suite_string = 0x03). It reuses the exact Ed25519 key: the VRF secret
 * scalar is clamp(SHA-512(seed)) and the VRF public key is the Ed25519 public
 * key. Curve/scalar arithmetic runs on a {@link CryptoBackend}.
 */
public final class Ecvrf {
    private Ecvrf() {}

    private static final byte SUITE = 0x03;
    private static final int PT_LEN = 32;
    private static final int C_LEN = 16;
    private static final int PROOF_LEN = PT_LEN + C_LEN + PT_LEN; // 80
    private static final byte[] IDENTITY = identity();

    private static byte[] identity() {
        byte[] id = new byte[32];
        id[0] = 1;
        return id;
    }

    /** A VRF proof pi = Gamma(32) || c(16) || s(32). */
    public record Proof(byte[] gamma, byte[] c, byte[] s) {
        public byte[] bytes() {
            return concat(gamma, c, s);
        }
    }

    /** Result of proving: the proof and the VRF output beta (64 bytes). */
    public record ProveResult(Proof proof, byte[] beta) {}

    /** Prove using the default backend. */
    public static ProveResult prove(VrfKey key, byte[] alpha) {
        return prove(key, alpha, Crypto.defaultBackend());
    }

    /** Verify using the default backend. */
    public static Optional<byte[]> verify(byte[] publicKey, byte[] alpha, byte[] pi) {
        return verify(publicKey, alpha, pi, Crypto.defaultBackend());
    }

    /** Prove: returns the proof pi and the VRF output beta (64 bytes). */
    public static ProveResult prove(VrfKey key, byte[] alpha, CryptoBackend backend) {
        byte[] seed = key.seed();
        byte[] x = Ed25519.secretScalar(seed, backend);
        byte[] y = backend.scalarmultBaseNoclamp(x); // public key Y = x*B
        byte[] h = encodeToCurve(y, alpha, backend);
        byte[] gamma = backend.scalarmultNoclamp(x, h).orElseThrow(() -> new IllegalStateException("Gamma = x*H failed"));
        byte[] k = nonce(seed, h, backend);
        byte[] u = backend.scalarmultBaseNoclamp(k);                              // U = k*B
        byte[] v = backend.scalarmultNoclamp(k, h).orElseThrow(() -> new IllegalStateException("V = k*H failed"));
        byte[] c = challenge(y, h, gamma, u, v, backend);                         // 16 bytes
        byte[] c32 = new byte[PT_LEN];
        System.arraycopy(c, 0, c32, 0, C_LEN);
        byte[] s = backend.scalarAdd(k, backend.scalarMul(c32, x));               // s = k + c*x mod L
        Proof proof = new Proof(gamma, c, s);
        return new ProveResult(proof, proofToHash(proof, backend));
    }

    /** Verify a proof against the public key and alpha. Present with beta if valid. */
    public static Optional<byte[]> verify(byte[] publicKey, byte[] alpha, byte[] pi, CryptoBackend backend) {
        Optional<Proof> decoded = decode(pi);
        if (decoded.isEmpty()) {
            return Optional.empty();
        }
        Proof proof = decoded.get();
        byte[] h = encodeToCurve(publicKey, alpha, backend);
        byte[] c32 = new byte[PT_LEN];
        System.arraycopy(proof.c, 0, c32, 0, C_LEN);

        byte[] sB = backend.scalarmultBaseNoclamp(proof.s);
        Optional<byte[]> cY = backend.scalarmultNoclamp(c32, publicKey);
        if (cY.isEmpty()) {
            return Optional.empty();
        }
        Optional<byte[]> u = backend.pointSub(sB, cY.get());                  // U = s*B - c*Y
        if (u.isEmpty()) {
            return Optional.empty();
        }
        Optional<byte[]> sH = backend.scalarmultNoclamp(proof.s, h);
        Optional<byte[]> cGamma = backend.scalarmultNoclamp(c32, proof.gamma);
        if (sH.isEmpty() || cGamma.isEmpty()) {
            return Optional.empty();
        }
        Optional<byte[]> v = backend.pointSub(sH.get(), cGamma.get());        // V = s*H - c*Gamma
        if (v.isEmpty()) {
            return Optional.empty();
        }
        byte[] cPrime = challenge(publicKey, h, proof.gamma, u.get(), v.get(), backend);
        if (!Arrays.equals(cPrime, proof.c)) {
            return Optional.empty();
        }
        return Optional.of(proofToHash(proof, backend));
    }

    /** proof_to_hash using the default backend. */
    public static byte[] proofToHash(Proof proof) {
        return proofToHash(proof, Crypto.defaultBackend());
    }

    /** proof_to_hash: beta = SHA-512(suite || 0x03 || point_to_string(8*Gamma) || 0x00). */
    public static byte[] proofToHash(Proof proof, CryptoBackend backend) {
        byte[] gamma8 = cofactorClear(proof.gamma, backend).orElseThrow(() -> new IllegalStateException("8*Gamma failed"));
        return backend.sha512(concat(new byte[]{SUITE, 0x03}, gamma8, new byte[]{0x00}));
    }

    public static Optional<Proof> decode(byte[] pi) {
        if (pi.length != PROOF_LEN) {
            return Optional.empty();
        }
        return Optional.of(new Proof(
                Arrays.copyOfRange(pi, 0, PT_LEN),
                Arrays.copyOfRange(pi, PT_LEN, PT_LEN + C_LEN),
                Arrays.copyOfRange(pi, PT_LEN + C_LEN, PROOF_LEN)));
    }

    // --- internals -----------------------------------------------------------

    /** ECVRF_encode_to_curve_try_and_increment (RFC 9381 §5.4.1.1). */
    private static byte[] encodeToCurve(byte[] pkString, byte[] alpha, CryptoBackend backend) {
        for (int ctr = 0; ctr <= 255; ctr++) {
            byte[] hashString = backend.sha512(
                    concat(new byte[]{SUITE, 0x01}, pkString, alpha, new byte[]{(byte) ctr}, new byte[]{0x00}));
            byte[] candidate = Arrays.copyOfRange(hashString, 0, PT_LEN); // string_to_point(first 32 bytes)
            Optional<byte[]> cleared = cofactorClear(candidate, backend);
            if (cleared.isPresent() && !Arrays.equals(cleared.get(), IDENTITY)) {
                return cleared.get();
            }
        }
        throw new IllegalStateException("encode_to_curve: no valid point found");
    }

    /** Multiply a compressed point by cofactor 8 via three doublings (on-curve check only). */
    private static Optional<byte[]> cofactorClear(byte[] p, CryptoBackend backend) {
        Optional<byte[]> p2 = backend.pointAdd(p, p);
        if (p2.isEmpty()) {
            return Optional.empty();
        }
        Optional<byte[]> p4 = backend.pointAdd(p2.get(), p2.get());
        if (p4.isEmpty()) {
            return Optional.empty();
        }
        return backend.pointAdd(p4.get(), p4.get());
    }

    /** ECVRF_nonce_generation_RFC8032 (RFC 9381 §5.4.2.2). */
    private static byte[] nonce(byte[] seed, byte[] hString, CryptoBackend backend) {
        byte[] kString = backend.sha512(concat(Ed25519.noncePrefix(seed, backend), hString));
        return backend.scalarReduce(kString);
    }

    /** ECVRF_challenge_generation (RFC 9381 §5.4.3): first 16 bytes. */
    private static byte[] challenge(byte[] y, byte[] h, byte[] gamma, byte[] u, byte[] v, CryptoBackend backend) {
        byte[] full = backend.sha512(concat(new byte[]{SUITE, 0x02}, y, h, gamma, u, v, new byte[]{0x00}));
        return Arrays.copyOfRange(full, 0, C_LEN);
    }

    private static byte[] concat(byte[]... parts) {
        int n = 0;
        for (byte[] p : parts) {
            n += p.length;
        }
        byte[] out = new byte[n];
        int pos = 0;
        for (byte[] p : parts) {
            System.arraycopy(p, 0, out, pos, p.length);
            pos += p.length;
        }
        return out;
    }
}
